File: haskell/app/Controllers/SalaController.hs
-- A modulo SalaController tem por intuito fornecer as funções para controlar e manipular os serviços de Sala no sistema.
module Controllers.SalaController where
-- Importações necessárias para o funcionamento do código.
import Repository
import Models.Sala
import Data.List (delete)
import Data.Time (UTCTime)
import Models.Usuario

-- Retorna todas as reservas que estão conflitando com a reserva passada
reservasConflitantes :: Reserva -> [Reserva] -> [Reserva]
reservasConflitantes reserva = filter (seReservaConflita reserva)

-- Verifica se duas reservas conflitam, ou seja, se os períodos de início e término se sobrepõem.
seReservaConflita :: Reserva -> Reserva -> Bool
seReservaConflita (Reserva _ inicio1 termino1) (Reserva _ inicio2 termino2) =
   (inicio2 < termino1 && inicio2 > inicio1) || (inicio1 < termino2 && inicio1 > inicio2)

-- Reserva uma sala
reservarSala :: Int -> Usuario -> UTCTime -> UTCTime -> IO (Either String Sala)
reservarSala numSalaId usuario inicio termino = do
    maybeSala <- fetchSala numSalaId
    case maybeSala of
        Nothing -> return $ Left "Sala não encontrada."
        Just sala -> do
            let reserva = Reserva (matriculaUsuario usuario) inicio termino

            let conflitos' = reservasConflitantes reserva (reservas sala)
            conflitos <- mapM (\r -> do
                maybeUser <- fetchUsuario (matricula r)
                -- Caso o usuario não existe por algum motivo, toma prioridade -1
                let prioridade = maybe (-1) prioridadeUsuario maybeUser
                return (prioridade, r)
                ) conflitos'
            
            let minhaPrioridade = prioridadeUsuario usuario
            -- Se o usuario tem prioridade sobre todas as reservas conflitantes
            let temPrioridade = all (\(p, _) -> minhaPrioridade > p) conflitos
            if not temPrioridade then
                return $ Left "Sala não disponível para reserva."
            else do
                -- Remove todas as reservas conflitantes
                let reservasAtualizadas = reserva : filter (`notElem` conflitos') (reservas sala)
                let salaAtualizada = sala { reservas = reservasAtualizadas }
                saveSala salaAtualizada
                return $ Right salaAtualizada

-- Função para cancelar uma reserva 
cancelarReserva :: Int -> Reserva -> IO (Either String Sala)
cancelarReserva numSalaId reservaParaCancelar = do
    maybeSala <- fetchSala numSalaId
    case maybeSala of
        Nothing -> return $ Left "Sala não encontrada."
        Just sala -> do
            if reservaParaCancelar `notElem` reservas sala
                then return $ Left "Reserva não existente."
                else do 
                    let reservasAtualizadas = delete reservaParaCancelar (reservas sala)
                    let salaAtualizada = sala { reservas = reservasAtualizadas }
                    saveSala salaAtualizada
                    return $ Right salaAtualizada

-- Lista todas as salas
listarSalas :: IO [Sala]
listarSalas = do
    fetchAllSalas

File: prolog/Controllers/SalaController.pl
:- module(sala_controller, [listar_salas/1, get_sala/2, reserva_conflitantes/3, se_reserva_conflita/2, sala_reservas_em_faixa/4, cancelar_reserva/3, reservar_sala/5]).

:- use_module(library(date)).
:- use_module('Repository.pl').
:- use_module('./Models/Usuario.pl').

% verifica se há reservas conflitantes com uma nova reserva nova. Recebe a reserva nova e a lista de reservas existentes e retorna a lista de reservas conflitantes.
reserva_conflitantes(NovaReserva, Reservas, Conflitantes) :-
    findall(R, (member(R, Reservas), se_reserva_conflita(NovaReserva, R)), Conflitantes).

% verifica se duas reservas são conflitantes, recebe duas reservas e retorna verdadeiro se as reservas forem conflitantes, se nao, e falso
se_reserva_conflita(Reserva1, Reserva2) :-
    Reserva1 = reserva(_, Inicio1, Termino1),
    Reserva2 = reserva(_, Inicio2, Termino2),
    not((Termino1 =< Inicio2; Inicio1 >= Termino2)).

% Lista todas as salas cadastradas.
listar_salas(Salas) :-
    fetch_all_salas(Salas).

% Retorna uma sala baseado no número de sala.
get_sala(NumeroSala, Resultado) :- repository:fetch_sala(NumeroSala, Resultado).

% retorna as reservas de uma sala que estão dentro de uma faixa de datas específicas, assim, recebe a sala, a data de início e a data de término e retorna uma lista de reservas que estão dentro da faixa de datas.
sala_reservas_em_faixa(Reservas, Inicio, Termino, ReservasEmFaixa) :-
    findall(Reserva, (member(Reserva, Reservas), se_reserva_conflita(Reserva, reserva(_, Inicio, Termino))), ReservasC),
    predsort(compare_reserva, ReservasC, ReservasEmFaixa).

% ordena as reservas por data de início de cada uma das duas reservas
compare_reserva(Order, reserva(_, I1, _), reserva(_, I2, _)) :-
    (   I1 @< I2 -> Order = '<'
    ;   I1 == I2 -> Order = '='
    ;   Order = '>'
    ).

% cancela uma reserva em uma sala, recebendo o número de sala, a reserva a ser cancelada e o resultado, se for cancelada com sucesso, ela retorna a sala atualizada.
cancelar_reserva(NumeroSala, ReservaParaCancelar, Resultado) :-
    (   get_sala(NumeroSala, Sala)
    ->  (   member(ReservaParaCancelar, Sala.reservas)
        ->  delete(Sala.reservas, ReservaParaCancelar, NovasReservas),
            NewSala = sala{numSala: Sala.numSala, nomeSala: Sala.nomeSala, reservas: NovasReservas},
            repository:save_sala(NewSala),
            Resultado = right(NewSala)
        ;   Resultado = left('Reserva não existente.')
        )
    ;   Resultado = left('Sala não encontrada.')
    ).

% reserva uma sala para um usuário específico em uma faixa de datas específica com base na sua prioridade. Recebe o número de sala, o usuário, a data de início e a data de término e retorna a sala atualizada com a reserva adicionada se a reserva foi realizada, se tiverem conflitos com outras reservas, ela retorna um erro.
reservar_sala(NumeroSala, Matricula, Inicio, Termino, Resultado) :-
    get_sala(NumeroSala, Sala),
    (   Reservas = Sala.reservas
    ->  NovaReserva = reserva(Matricula, Inicio, Termino),
        reserva_conflitantes(NovaReserva, Reservas, Conflitantes),
        findall((PriConflitante, ResConflitante), (
            member(ResConflitante, Conflitantes),
            ResConflitante = reserva(MatriculaRes, _, _),
            fetch_usuario(MatriculaRes, UserConflit),
            prioridade_usuario(UserConflit, PriConflitante)
        ), Conflitos),
        fetch_usuario(Matricula, U),
        prioridade_usuario(U, MinhaPrioridade),
        (   forall(member((PriConflitante, ResConflitante), Conflitos), MinhaPrioridade > PriConflitante)
        ->  subtract(Reservas, Conflitantes, ReservasAtualizadas),
            append([NovaReserva], ReservasAtualizadas, NovasReservas),
            NewSala = Sala.put(_{reservas: NovasReservas}),
            repository:save_sala(NewSala),
            Resultado = right(NewSala)
        ;   Resultado = left('Sala não disponível para reserva.')
        )
    ;   Resultado = left('Sala não encontrada.')
    ).